File: bindings/dotnet/src/Aster/Interceptors/CircuitBreakerInterceptor.cs
namespace Aster.Interceptors;

/// <summary>
/// Simple CLOSED -> OPEN -> HALF_OPEN circuit breaker.
/// </summary>
public class CircuitBreakerInterceptor : IInterceptor
{
    public const string StateClosed = "closed";
    public const string StateOpen = "open";
    public const string StateHalfOpen = "half_open";

    private static readonly HashSet<StatusCode> TrippingCodes = new()
    {
        StatusCode.Unavailable,
        StatusCode.Internal,
        StatusCode.Unknown,
    };

    /// <summary>Number of consecutive failures before opening.</summary>
    public int FailureThreshold { get; }

    /// <summary>Seconds to wait before transitioning from Open to HalfOpen.</summary>
    public double RecoveryTimeout { get; }

    /// <summary>Max calls allowed in HalfOpen state before re-evaluating.</summary>
    public int HalfOpenMaxCalls { get; }

    /// <summary>Current circuit state.</summary>
    public string State { get; private set; } = StateClosed;

    /// <summary>Consecutive failure count.</summary>
    public int FailureCount { get; private set; }

    private long _openedAtTicks;
    private int _halfOpenCalls;

    public CircuitBreakerInterceptor(
        int failureThreshold = 3,
        double recoveryTimeout = 5.0,
        int halfOpenMaxCalls = 1)
    {
        FailureThreshold = failureThreshold;
        RecoveryTimeout = recoveryTimeout;
        HalfOpenMaxCalls = halfOpenMaxCalls;
    }

    public object OnRequest(CallContext ctx, object request)
    {
        BeforeCall();
        return request;
    }

    public RpcError? OnError(CallContext ctx, RpcError error)
    {
        RecordFailure(error);
        return error;
    }

    public object OnResponse(CallContext ctx, object response)
    {
        RecordSuccess();
        return response;
    }

    /// <summary>
    /// Check circuit state before making a call. Throws RpcError if circuit is open.
    /// </summary>
    public void BeforeCall()
    {
        long now = Environment.TickCount64;

        if (State == StateOpen)
        {
            double elapsed = (now - _openedAtTicks) / 1000.0;
            if (elapsed >= RecoveryTimeout)
            {
                State = StateHalfOpen;
                _halfOpenCalls = 0;
            }
            else
            {
                throw new RpcError(StatusCode.Unavailable, "circuit breaker is open");
            }
        }

        if (State == StateHalfOpen)
        {
            if (_halfOpenCalls >= HalfOpenMaxCalls)
            {
                throw new RpcError(StatusCode.Unavailable, "circuit breaker is half-open");
            }
            _halfOpenCalls++;
        }
    }

    /// <summary>Record a successful call -- resets the circuit to closed.</summary>
    public void RecordSuccess()
    {
        FailureCount = 0;
        _halfOpenCalls = 0;
        State = StateClosed;
    }

    /// <summary>Record a failed call -- may trip the circuit to open.</summary>
    public void RecordFailure(RpcError error)
    {
        if (!TrippingCodes.Contains(error.Code))
            return;

        if (State == StateHalfOpen)
        {
            State = StateOpen;
            _openedAtTicks = Environment.TickCount64;
            _halfOpenCalls = 0;
            return;
        }

        FailureCount++;
        if (FailureCount >= FailureThreshold)
        {
            State = StateOpen;
            _openedAtTicks = Environment.TickCount64;
        }
    }
}

File: bindings/java/aster-runtime/src/main/java/site/aster/interceptors/CircuitBreakerInterceptor.java
package site.aster.interceptors;

import java.util.Set;

/**
 * Simple CLOSED -> OPEN -> HALF_OPEN circuit breaker.
 *
 * <p>Tracks consecutive failures and opens the circuit when the threshold is reached. After a
 * recovery timeout, the circuit moves to half-open and allows a limited number of probe calls.
 * Success resets to closed; failure reopens immediately.
 */
public final class CircuitBreakerInterceptor implements Interceptor {

  /** Circuit breaker states. */
  public enum State {
    CLOSED,
    OPEN,
    HALF_OPEN
  }

  private static final Set<StatusCode> TRIP_CODES =
      Set.of(StatusCode.UNAVAILABLE, StatusCode.INTERNAL, StatusCode.UNKNOWN);

  private final int failureThreshold;
  private final double recoveryTimeoutSecs;
  private final int halfOpenMaxCalls;

  private volatile State state = State.CLOSED;
  private int failureCount;
  private long openedAtNanos;
  private int halfOpenCalls;

  /**
   * Creates a circuit breaker interceptor.
   *
   * @param failureThreshold consecutive failures before opening (default 3)
   * @param recoveryTimeoutSecs seconds to wait before transitioning to half-open (default 5.0)
   * @param halfOpenMaxCalls max probe calls in half-open state (default 1)
   */
  public CircuitBreakerInterceptor(
      int failureThreshold, double recoveryTimeoutSecs, int halfOpenMaxCalls) {
    this.failureThreshold = failureThreshold;
    this.recoveryTimeoutSecs = recoveryTimeoutSecs;
    this.halfOpenMaxCalls = halfOpenMaxCalls;
  }

  /** Creates a circuit breaker with default settings (threshold=3, recovery=5s, halfOpen=1). */
  public CircuitBreakerInterceptor() {
    this(3, 5.0, 1);
  }

  /** Returns the current circuit breaker state. */
  public State state() {
    return state;
  }

  @Override
  public synchronized Object onRequest(CallContext ctx, Object request) {
    long now = System.nanoTime();

    if (state == State.OPEN) {
      double elapsedSecs = (now - openedAtNanos) / 1_000_000_000.0;
      if (elapsedSecs >= recoveryTimeoutSecs) {
        state = State.HALF_OPEN;
        halfOpenCalls = 0;
      } else {
        throw new RpcError(StatusCode.UNAVAILABLE, "circuit breaker is open");
      }
    }

    if (state == State.HALF_OPEN) {
      if (halfOpenCalls >= halfOpenMaxCalls) {
        throw new RpcError(StatusCode.UNAVAILABLE, "circuit breaker is half-open");
      }
      halfOpenCalls++;
    }

    return request;
  }

  @Override
  public synchronized Object onResponse(CallContext ctx, Object response) {
    // Success -- reset to closed
    failureCount = 0;
    halfOpenCalls = 0;
    state = State.CLOSED;
    return response;
  }

  @Override
  public synchronized RpcError onError(CallContext ctx, RpcError error) {
    if (!TRIP_CODES.contains(error.code())) {
      return error;
    }

    if (state == State.HALF_OPEN) {
      state = State.OPEN;
      openedAtNanos = System.nanoTime();
      halfOpenCalls = 0;
      return error;
    }

    failureCount++;
    if (failureCount >= failureThreshold) {
      state = State.OPEN;
      openedAtNanos = System.nanoTime();
    }

    return error;
  }
}
